### PR TITLE
Add line count check to avoid out of range

### PR DESCRIPTION
**Pull request by valterc**
IndexOutOfRangeException is thrown in [StringLineGroup](https://github.com/xoofx/markdig/blob/master/src/Markdig/Helpers/StringLineGroup.cs) when calling Iterator.NextChar 
https://github.com/xoofx/markdig/blob/8155a1e3d657dbe17c14e5612791b4a7d7f3da68/src/Markdig/Helpers/StringLineGroup.cs#L276

The exception occurs when trying to advance to the next line when the StringLineGroup is at capacity (capacity == line_count) 

https://github.com/xoofx/markdig/blob/8155a1e3d657dbe17c14e5612791b4a7d7f3da68/src/Markdig/Helpers/StringLineGroup.cs#L339

This exception is kind of an edge case and does not happen often because `LeafBlock` and other parts that use the StringLineGroup provision some extra capacity in advance and when capacity is needed it is increased by a factor of 2, so it's not very common that the needed capacity matches the line count.

### CODE REVIEW

**Review by MihaZupan — COMMENTED** [review]
Did you hit this edge-case with any real-world input while parsing?

Looks like some tests aren't happy with this change, mind looking at those?

**Comment by valterc** [comment]
My bad, I didn't really run all the tests locally.

I did hit this case when I was parsing a yaml front matter which contained exactly 8 lines inside of it. Looking through the code, `LeafBlock` reserves 4 lines up-front:
https://github.com/xoofx/markdig/blob/8155a1e3d657dbe17c14e5612791b4a7d7f3da68/src/Markdig/Syntax/LeafBlock.cs#L82
And then the `StringLineGroup` increases the capacity to 8, when using the  `CharIterator` this bug comes up.

In my case I was building a subclass of `TextReader` based on the `CharIterator` (avoiding allocations) to use in some other frameworks.

I've added a sort of end-to-end test in `TestYamlFrontMatterExtension`, let me know if that is OK or should I add in the RoundTrip tests. Without the fix, the test fails with exception:
image

Example markdown that breaks the `CharIterator`:

```markdown
---
key1: value1
key2: value2
key3: value3
key4: value4
key5: value5
key6: value6
key7: value7
key8: value8
---

# Content
```

File: src/Markdig.Tests/TestStringLineGroup.cs [hunk 1]
```diff
@@ -0,0 +1,55 @@
+using Markdig.Helpers;
+using System.Text;
+
+namespace Markdig.Tests;
+
+[TestFixture]
+public class TestStringLineGroup
```
[the remark this answers is not in the document]
Comment (valterc) [reply src/Markdig.Tests/TestStringLineGroup.cs]: Done.
